cryptoprotocol/src/main/java/org/cryptimeleon/incentivesystem/cryptoprotocol/proof/SpendDeductZkp.java

```diff
@@ -0,0 +1,164 @@
+        }
+
+        // v >= k (I have more points than required)
+        // We prove that v-k\in[0,eskDecBase^{maxPointBasePower+1}-1]
+        builder.addSubprotocol("v>=k", new SmallerThanPowerFragment(vVar.sub(commonInput.k), pp.getEskDecBase().getInteger().intValue(), pp.getMaxPointBasePower(), pp.getEskBaseSetMembershipPublicParameters()));
+
+        // ctrace=(w^{r_i} ,{w^{r_i}}^{esk}*w^{esk^*_{usr,i}}) for all i\in[p]
+        for (int i = 0; i < pp.getNumEskDigits(); i++) {
+            builder.addSubprotocol("ctrace0" + i, new LinearStatementFragment(w.pow(rVector.get(i)).isEqualTo(commonInput.ctrace0.get(i))));
+            builder.addSubprotocol("ctrace1" + i, new LinearStatementFragment(w.pow(eskDecVarVector.get(i)).op(commonInput.ctrace0.get(i).pow(eskVar)).isEqualTo(commonInput.ctrace1.get(i))));
+        }
+
+        return builder.build();
+    }
+
+    @Override
+    protected SendThenDelegateFragment.ProverSpec provideProverSpecWithNoSendFirst(CommonInput pCommonInput, SecretInput pSecretInput, SendThenDelegateFragment.ProverSpecBuilder builder) {
+        var secretInput = (SpendDeductZkpWitnessInput) pSecretInput;
+
+        // Add variables to witness
+        builder.putWitnessValue("esk", secretInput.esk);
+        builder.putWitnessValue("usk", secretInput.usk);
+        builder.putWitnessValue("dsrnd0", secretInput.dsrnd0);
+        builder.putWitnessValue("dsrnd1", secretInput.dsrnd1);
+        builder.putWitnessValue("dsrndStar0", secretInput.dsrndStar0);
+        builder.putWitnessValue("dsrndStar1", secretInput.dsrndStar1);
+        builder.putWitnessValue("v", secretInput.v);
+        builder.putWitnessValue("z", secretInput.z);
+        builder.putWitnessValue("t", secretInput.t);
+        builder.putWitnessValue("zStar", secretInput.zStar);
+        builder.putWitnessValue("tStar", secretInput.tStar);
+        builder.putWitnessValue("uStarInverse", secretInput.uStar.inv());
+
+        for (int i = 0; i < pp.getNumEskDigits(); i++) {
+            builder.putWitnessValue("eskStarUserDec_" + i, (Zn.ZnElement) secretInput.eskStarUserDec.get(i));
+            builder.putWitnessValue("r_" + i, (Zn.ZnElement) secretInput.rVector.get(i));
+        }
+
+        // Some asserts that might be useful for debugging:
+        // assert pp.getNumEskDigits() == secretInput.eskStarUserDec.length();
+        // assert secretInput.eskStarUser.equals(secretInput.eskStarUserDec.map((integer, znElement) -> znElement.mul(pp.getEskDecBase().pow(BigInteger.valueOf(integer)))).reduce(Zn.ZnElement::add));
+        // assert commonInput.ctrace0.equals(pp.getW().pow(secretInput.rVector));
+        // assert commonInput.ctrace1.equals(commonInput.ctrace0.pow(secretInput.esk).op(pp.getW().pow(secretInput.eskStarUserDec)));
+
+        return builder.build();
+    }
+
+    @Override
+    public BigInteger getChallengeSpaceSize() {
+        return groupG1.size();
+    }
+}
```

Comment:
Again, what was modified here since last review? I remember seeing this class before but it is still marked as entirely changed.

Reply:
I think nothing